smack-websocket/src/main/java/org/jivesoftware/smack/websocket/XmppWebsocketTransportModule.java

```diff
@@ -0,0 +1,326 @@
+
+                    // Generate Result for explicitly configured endpoint.
+                    Result manualResult = new Result(Arrays.asList(providedEndpoint), null);
+
+                    LookupConnectionEndpointsResult endpointsResult = new DiscoveredWebsocketEndpoints(manualResult);
+
+                    websocketEndpointsLookupFuture.setResult(endpointsResult);
+                } else {
+                    // Fetch remote endpoints.
+                    Result xep0156result = WebsocketRemoteConnectionEndpointFinder.find(connectionInternal);
+
+                    List<WebsocketRemoteConnectionEndpoint> discoveredEndpoints = xep0156result.discoveredRemoteConnectionEndpoints;
+
+                    int discoveredEndpointCount = xep0156result.discoveredRemoteConnectionEndpoints.size();
+                    List<WebsocketRemoteConnectionEndpoint> endpoints = new ArrayList<>(discoveredEndpointCount);
+
+                    // Add remote discovered endpoints.
+                    endpoints.addAll(discoveredEndpoints);
+
+                    // Generate result considering both manual and fetched endpoints.
+                    Result finalResult = new Result(endpoints, xep0156result.getLookupFailures());
+
+                    LookupConnectionEndpointsResult endpointsResult = new DiscoveredWebsocketEndpoints(finalResult);
+
+                    websocketEndpointsLookupFuture.setResult(endpointsResult);
+                }
+            });
+
+            return Collections.singletonList(websocketEndpointsLookupFuture);
+        }
+
+        @Override
+        protected void loadConnectionEndpoints(LookupConnectionEndpointsSuccess lookupConnectionEndpointsSuccess) {
+            discoveredWebsocketEndpoints = (DiscoveredWebsocketEndpoints) lookupConnectionEndpointsSuccess;
+        }
+
+        @Override
+        protected void afterFiltersClosed() {
+        }
+
+        @Override
+        protected void disconnect() {
+            websocket.disconnect(1000, "Websocket closed normally");
+        }
+
+        @Override
+        protected void notifyAboutNewOutgoingElements() {
+            Queue<TopLevelStreamElement> outgoingElementsQueue = connectionInternal.outgoingElementsQueue;
+            asyncButOrderedOutgoingElementsQueue.performAsyncButOrdered(outgoingElementsQueue, () -> {
+                // Once new outgoingElement is notified, send the top level stream element obtained by polling.
+                TopLevelStreamElement topLevelStreamElement = outgoingElementsQueue.poll();
+                websocket.send(topLevelStreamElement);
+            });
+        }
+
+        @Override
+        public SSLSession getSslSession() {
+           return websocket.getSSLSession();
+        }
+
+        @Override
+        public boolean isTransportSecured() {
+            return websocket.isConnectionSecure();
+        }
+
+        @Override
+        public boolean isConnected() {
+            return connectionInternal.connection.isConnected();
```

Comment:
That looks like it could easily cause an infinite recursion. This method is meant to ask the underlying transport if he believes to be connected. Unfortunately, okhttp does not provide such an API, so we have to remember the state ourselves.